_source/_posts/2024-02-27-spring-security-pkce.md

```diff
@@ -0,0 +1,458 @@
+
+The diagram is a simplified sequence of the Authorization Code Flow, where the User Agent (the browser) redirections are not shown.
+
+When calling the Token Endpoint, the client application must authenticate itself. The protocol provides several ways for [client authentication](https://openid.net/specs/openid-connect-core-1_0.html#ClientAuthentication):
+
+- `client_secret_basic`: The client was assigned a secret (confidential client) and authenticates using HTTP basic authentication scheme.
+- `client_secret_post`: The client was assigned a secret and authenticates by including the client credentials in the request body.
+- `client_secret_jwt`: The client was assigned a secret and authenticates using a JWT signed with part of the secret as a shared key.
+- `private_key_jwt`: The client has registered a public key and authenticates using a JWT signed with the private key.
+- `none`: The client does not authenticate itself at the Token Endpoint, because it is a public client.
+
+As browser and mobile applications _cannot_ hold credentials securely and therefore cannot identify themselves using a client secret, [PKCE](https://www.rfc-editor.org/rfc/rfc7636) was created for extending the OAuth 2.0 Authorization Code Flow, adding a dynamically created cryptographically random key called "code verifier". This extension was created to mitigate the [authorization code interception attack](https://www.rfc-editor.org/rfc/rfc7636#section-1).
+
+The modified flow has the following steps:
+
+{% img blog/spring-security-pkce/auth0-authorization-code-pkce.png alt:"Authorization Code Grant" width:"850" %}{: .center-image }
+
+1. The Client creates and records a secret named the "code_verifier" and derives a transformed version referred to as the "code_challenge", which is sent in the OAuth 2.0 Authorization Request along with the transformation method.
+2. The Authorization Endpoint responds as usual but records the "code_challenge" and the transformation method.
+3. The Client then sends the authorization code in the Access Token Request as usual but includes the "code_verifier" secret generated in the first step.
+4. The Authorization Server transforms "code_verifier" and compares it to the recorded "code_challenge". Access is denied if they are not equal.
+
+The latest [Security BCP](https://www.ietf.org/archive/id/draft-ietf-oauth-security-topics-24.html) states that PKCE should be enabled for all types of clients, public and confidential (browser-based applications, mobile applications, native applications, and secure server applications).
+
+## Spring Security for Authorization Code Flow
+
+You can experiment with how to configure the authorization code flow configuration by creating a simple Spring Boot web application, following the step-by-step guide in the following sections.
+
+
+### Create a simple Spring Boot application
+
+With Spring Initializr and curl, create Spring Boot project:
+
+```shell
+curl -v "https://start.spring.io/starter.zip?\
```

Comment:
Please use this command instead. Most of the args are not required as they are the default
```
curl -G https://start.spring.io/starter.tgz \
  -d dependencies=web,okta,thymeleaf \
  -d baseDir=passkey-demo \
 | tar -xzvf -
```

Reply:
Partial update, I keep non-required default args as documentation